### PR TITLE
Get block by closest timestamp

### PR DESCRIPTION
**Pull request by bronzelle-cw**
### **PR Type**
Enhancement

___

### **Description**
- Implement 'read_block_by_timestamp' functionality

- Add logic to find closest block by timestamp

- Update RPC server to use new timestamp-based block retrieval

- Add e2e tests for timestamp-based block fetching

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["RPC Server"] --> B["StratusStorage"]
  B --> C["RocksPermanentStorage"]
  C --> D["RocksStorageState"]
  D --> E["Block Retrieval"]
  E --> F["Timestamp-based Search"]
  G["E2E Tests"] --> A
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>rpc_server.rs</strong><dd><code>Update RPC server for timestamp-based block retrieval</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/rpc/rpc_server.rs

<ul><li>Updated <code>stratus_get_block_by_timestamp</code> to use new <br><code>read_block_by_timestamp</code> method<br> <li> Removed <code>BlockFilter::Timestamp</code> usage<br> <li> Improved tracing for timestamp-based block retrieval</ul>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/2389/files#diff-835ba255c9a5c1fe0e13285bc39e058e1c74422e5e03ebba483c9d8a15c45405">+4/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>rocks_permanent.rs</strong><dd><code>Add timestamp-based block retrieval to RocksPermanentStorage</code></dd></summary>
<hr>

src/eth/storage/permanent/rocks/rocks_permanent.rs

<ul><li>Added <code>read_block_by_timestamp</code> method to <code>RocksPermanentStorage</code><br> <li> Implemented error handling and tracing for the new method</ul>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/2389/files#diff-d582062cac05f2b057f597f872d3f0415f29f579876306b2d9289c278c3bd3fd">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>rocks_state.rs</strong><dd><code>Implement core logic for timestamp-based block retrieval</code>&nbsp; </dd></summary>
<hr>

src/eth/storage/permanent/rocks/rocks_state.rs

<ul><li>Implemented <code>read_block_by_timestamp</code> method in <code>RocksStorageState</code><br> <li> Added logic to find the closest block by timestamp<br> <li> Handled edge cases for timestamp-based block retrieval</ul>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/2389/files#diff-9d38c98aa2d77c1665d2e2e11a0abc4787424d1b41a610d2fc1c4ea0311014a9">+46/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>stratus_storage.rs</strong><dd><code>Add timestamp-based block retrieval to StratusStorage</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/storage/stratus_storage.rs

<ul><li>Added <code>read_block_by_timestamp</code> method to <code>StratusStorage</code><br> <li> Implemented tracing and metrics for the new method</ul>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/2389/files#diff-3b2e36e47959decdfe26b2c4fab90b41f9242bfdad29cc43b2377ee851f4f40a">+14/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>e2e-json-rpc.test.ts</strong><dd><code>Add e2e tests for timestamp-based block retrieval</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/test/external/e2e-json-rpc.test.ts

<ul><li>Added new test cases for <code>stratus_getBlockByTimestamp</code><br> <li> Tested various scenarios including exact, closest, and edge cases<br> <li> Verified correct block retrieval based on timestamps</ul>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/2389/files#diff-f5d1f06c4777826f9cbdcb4372d15c42f354177c83df65f3324cfd607da16004">+30/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: src/eth/storage/permanent/rocks/rocks_state.rs [hunk 1]
```diff
@@ -422,6 +423,51 @@ impl RocksStorageState {
         block.map(|block_option| block_option.map(|block| block.into_inner().into()))
     }
 
+    pub fn read_block_by_timestamp(&self, target: UnixTime) -> Result<Option<Block>> {
+        tracing::debug!(timestamp = %target, "reading block by timestamp");
+
+        let timestamp_key: UnixTimeRocksdb = target.into();
+
+        let load_block = |block_number: &CfBlocksByHashValue| -> Result<Option<Block>> {
```
Comment (carneiro-cw) [inline src/eth/storage/permanent/rocks/rocks_state.rs]: Instead of using this closure you can just find the number and get+convert on the last line of the function

File: src/eth/storage/permanent/rocks/rocks_state.rs [hunk 2]
```diff
@@ -422,6 +423,51 @@ impl RocksStorageState {
...
+        if let Some(block_number) = self.blocks_by_timestamp.get(&timestamp_key)? {
+            return load_block(&block_number);
+        }
+
+        let target_value = *target;
+        let next_entry = self.blocks_by_timestamp.seek(timestamp_key)?.filter(|(ts, _)| ts.0 >= target_value);
+
+        let prev_iter = self.blocks_by_timestamp.iter_from(timestamp_key, Direction::Reverse)?;
+        let mut prev_entry = None;
+        for entry in prev_iter {
```
Comment (carneiro-cw) [inline src/eth/storage/permanent/rocks/rocks_state.rs]: You only need the first value (call `.next()`)

File: src/eth/storage/permanent/rocks/rocks_state.rs [hunk 3]
```diff
@@ -422,6 +423,51 @@ impl RocksStorageState {
...
+
+        let timestamp_key: UnixTimeRocksdb = target.into();
+
+        let load_block = |block_number: &CfBlocksByHashValue| -> Result<Option<Block>> {
+            self.blocks_by_number
+                .get(block_number)
+                .map(|block_opt| block_opt.map(|block| block.into_inner().into()))
+        };
+
+        if let Some(block_number) = self.blocks_by_timestamp.get(&timestamp_key)? {
```
[the remark this answers is not in the document]
Comment (bronzelle-cw) [reply src/eth/storage/permanent/rocks/rocks_state.rs]: I thought it could be faster getting it directly if the timestamp was exact, but you're right. The iterator is too short, so it's ok just use it.
